votingworks/vxsuite
# Apply New Printing Approach to HMPBs

## Overview
Starts #2759 and closes #2760. 

Extends the new printing approach - in which all prints are rendered outside the app body momentarily before being printed - to printing hand-marked paper ballots in VxAdmin. I wanted to start with hand-marked paper ballots because, as our only component which needs post-processing before it can be printed - I wanted to confirm its feasibility before applying the new approach across VxAdmin.

The HMPB component now behaves more like a normal component in that it renders to _itself_, not a specified `targetElement`. This way, we don't have strange ordering problems.

The hand-marked paper ballot component is used in five places in the application and this PR is separated into five commits, each representing one of those places. Highly recommend commit-by-commit review.

### 1 - Ballot Screen
As a first step, print ballots from the ballot screen with the new approach. This involved the change to the HMPB component itself. I found that `PagedJs` cannot work within a `display: none;` container because it uses `getBoundingClientRect`. In places where we've been wrapping HMPB's in `print-only` containers in the past, that container has been completely meaningless because the component is actually rendered to the `targetElement`. But `PagedJs` does not show the chunked (paged) content on screen. 

This change also required making a new print button (called `NewPrintButton` for now) that doesn't do anything to render an element. 

### 2 - Test Deck Screen
Removed the `targetElement` logic from the test deck screen, because the updated HMPB component renders in order like a normal component. Using the new printing approach for the test decks will be in a separate PR.

### 3 - Print All Ballots Modal
Applied the new approach to the Print All Ballots modal. It's a good example of how the new printing approach allows a more sequential, clear print sequence that doesn't require coordinating component state and rendering (although we often will want to coordinate component state to show printing status updates).

### 4 - Save All Ballot PDFs Modal
I was hoping to avoid touching #2760 in this PR, but `PagedJs` doesn't play nice within `ReactModal` and it needed to be done. I extended the `printElement` utility to have an alternate signature that returns PDF data. There may have been advantages to just sharing an exported signature and determining return type based on an `option` value - for example, the existing `expectPrint` testing utility could then just be used for PDFs. But they are different enough that reads better to separate them. As such, there is no `expectPrint` equivalent for `printElementToPDF` yet. That wasn't a problem for this PR because we have no testing around the content of documents printed to PDF. Although we could, and we should, just not here.

### 5 - Export Ballot Package
Use new approach for the PDFs in the ballot package.

## Testing Plan
Since this was mostly a refactor, goal was to limit automated testing changes. On manual testing, I tested each of the 5 places where HMPBs are generated with `kiosk-browser` on hardware:

1. Confirm ballot printed through `BallotScreen` scans
2. Confirm ballot printed through test deck scans
3. Confirm ballot printed through "Print All" modal scans
4. Confirm that ballot printed from PDF export scans
5. Confirm that we can successfully configure, and then scan, with ballot package

## Checklist
- [ ] ~~I have added [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging) where appropriate to any new user actions, system updates such as file reads or storage writes, or errors introduced.~~
- [ ] I have added JSDoc comments to any newly introduced exports

- [ ] ~~I have added a screenshot and/or video to this PR to demo the change~~
- [ ] ~~I have added the "user_facing_change" label to this PR to automate an announcement in #machine-product-updates~~
